### PR TITLE
Fix tar prefix header

### LINKED ISSUE #33958: failed to register layer: Error processing tar file(exit status 1): chtimes /rc/lib: no such file or directory
**Description**
On docker 17.06 cannot pull some images:
```failed to register layer: Error processing tar file(exit status 1): chtimes /rc/lib: no such file or directory```

**Steps to reproduce the issue:**
1. Upgrade docker to 17.06 or fresh install
2. docker pull zigmund/pub:r8

**Describe the results you received:**
```
zigmund@docker-host:~$  docker pull zigmund/pub:r8
r8: Pulling from zigmund/pub
6f821164d5b7: Pull complete 
e8580501e1fa: Pull complete 
90c325f7fe6a: Pull complete 
2f4daf9f9ce2: Pull complete 
4adbfe2315cf: Extracting [==================================================>]  124.3MB/124.3MB
6bf8f972dbfc: Download complete 
4f4fc94abdfa: Download complete 
495fc8389d76: Download complete 
473b636bf8bc: Download complete 
f53fa98eccc2: Download complete 
fab3c5f07ffc: Download complete 
356a7a6a20cf: Download complete 
failed to register layer: Error processing tar file(exit status 1): chtimes /rc/lib: no such file or directory
```

**Describe the results you expected:**
Normally pulled image.

**Additional information you deem important (e.g. issue happens only occasionally):**
Tried aufs and ovelay2 storage drivers - same result.
Image built on docker 17.05. Only some images cannot be pulled. All that problem images can be pulled normally on docker 17.03 - 17.05.
I pushed example to zigmund/pub:r8 for testing.
As temporary solution - rebuild problem image.

**Output of `docker version`:**
```
Client:
 Version:      17.06.0-ce
 API version:  1.30
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:23:31 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.06.0-ce
 API version:  1.30 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   02c1d87
 Built:        Fri Jun 23 21:19:04 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 17.06.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: gelf
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: cfb82a876ecc11b5ca0977d1733adbe58599088a
runc version: 2d41c047c83e09a6d61d464906feb2a2f3c52aa4
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-83-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.796GiB
Name: docker-w5
ID: SGAF:WNLR:GEHQ:3KAF:Q36I:SEV2:KQWC:BVHC:GPCX:LTRN:VQ5F:MIJF
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Ubuntu 16.04.2 LTS 4.4.0-83-generic. Virtual machine on Proxmox 4.4.

### LINKED ISSUE #34092: Difference in image extraction between Docker CE 17.05 and 17.06
**Description**

Image extraction between Docker CE 17.05 and 17.06 results in different root filesystems. I am able to reproduce the issue with both aufs and overlay2 storage drivers.

I ran into this issue with an image from a private repository. If the information that follows isn't enough to identify the issue, please let me know and I will attempt to create a public image that reproduces the behaviour described.

**Steps to reproduce the issue:**

Using Docker 17.06 CE

1. `docker-machine create --virtualbox-boot2docker-url "https://github.com/boot2docker/boot2docker/releases/download/v17.06.0-ce/boot2docker.iso" -d virtualbox --engine-insecure-registry internal_repo 1706ce`
2. `eval $(docker-machine env 1706ce)`
3. `docker run --rm -ti internal_repo/image ls /`

**Describe the results you received:**

```
#
0001-gyp-always-install-into-PRODUCT_DIR.patch
0002-gyp-apply-https-codereview.chromium.org-11361103.patch
0003-gyp-don-t-use-links-at-all-just-copy-the-files-inste.patch
2015-01-30.md
@@iterator
AUTHORS
AUTHORS.md
BufferList.js
CHANGELOG.md
CHANGES
CHANGES.md
CONTRIBUTING.md
COPYING.txt
CodecMalformedException.java

<~1000 output lines truncated>
```

**Describe the results you expected:**

Using Docker 17.05 CE

1. `docker-machine create --virtualbox-boot2docker-url "https://github.com/boot2docker/boot2docker/releases/download/v17.05.0-ce/boot2docker.iso" -d virtualbox --engine-insecure-registry internal_repo 1705ce`
2. `eval $(docker-machine env 1705ce)`
3. `docker run --rm -ti internal_repo/image ls /`

result:

```
INFORMIXTMP  artifacts	bin  boot  database  dev  etc  home  lib  lib64  media	mnt  nohup.out	nonexistent  opt  proc	root  run  sbin  selinux  shared  srv  sys  tmp  usr  var
```

**Additional information**

The filesystem in the 17.06 case is not completely flattened. It does contain the directories shown in the 17.05 output, and they do contain files.

An excerpt follows showing the result of comparing file checksums/locations between the 17.06 and 17.05 cases:

```
< 00564742e3c70905fdb4d3318b01c827  /node_modules/es5-ext/.lint
---
> 00564742e3c70905fdb4d3318b01c827  /usr/lib/node_modules/npm/node_modules/node-gyp/node_modules/path-array/node_modules/array-index/node_modules/es6-symbol/node_modules/es5-ext/.lint
145a146,147
> 00564742e3c70905fdb4d3318b01c827  /usr/local/lib/node_modules/npm/node_modules/node-gyp/node_modules/path-array/node_modules/array-index/node_modules/es6-symbol/node_modules/es5-ext/.lint
> 00564742e3c70905fdb4d3318b01c827  /usr/local/n/versions/node/7.0.0/lib/node_modules/npm/node_modules/node-gyp/node_modules/path-array/node_modules/array-index/node_modules/es6-symbol/node_modules/es5-ext/.lint
322a325,327
> 00ac2be339a766e2df37641d0350c63e  /usr/lib/node_modules/npm/node_modules/read-package-json/node_modules/glob/node_modules/path-is-absolute/package.json
> 00ac2be339a766e2df37641d0350c63e  /usr/local/lib/node_modules/npm/node_modules/read-package-json/node_modules/glob/node_modules/path-is-absolute/package.json
> 00ac2be339a766e2df37641d0350c63e  /usr/local/n/versions/node/7.0.0/lib/node_modules/npm/node_modules/read-package-json/node_modules/glob/node_modules/path-is-absolute/package.json
492a498,500
> 01028d57163f4901221a7a457774e202  /usr/lib/node_modules/npm/node_modules/init-package-json/node_modules/glob/node_modules/path-is-absolute/package.json
> 01028d57163f4901221a7a457774e202  /usr/local/lib/node_modules/npm/node_modules/init-package-json/node_modules/glob/node_modules/path-is-absolute/package.json
> 01028d57163f4901221a7a457774e202  /usr/local/n/versions/node/7.0.0/lib/node_modules/npm/node_modules/init-package-json/node_modules/glob/node_modules/path-is-absolute/package.json
493a502
> 0103318a939e4c2198689f6f143e556d  /usr/lib/node_modules/npm/node_modules/request/node_modules/http-signature/node_modules/sshpk/lib/formats/x509.js
495c504,505
< 0103318a939e4c2198689f6f143e556d  /x509.js
---

```
It can be seen that several files are missing.

`/usr/lib/node_modules/npm/node_modules/node-gyp/node_modules/path-array/node_modules/array-index/node_modules/es6-symbol/node_modules/es5-ext/.lint` under 17.05 moved to `/node_modules/es5-ext/.lint` in 17.06 and `/usr/lib/node_modules/npm/node_modules/request/node_modules/http-signature/node_modules/sshpk/lib/formats/x509.js` in 17.05 moved to `/x509.js` in 17.06. File contents are unaffected.

Here is the output of `docker inspect internal_repo/image` for the 17.05 case:

```
[
    {
        "Id": "sha256:a126cdbff382390a94be039424f1db5f6022ecb4444f6c4ea78853cc774dee12",
        "RepoTags": [
            "internal_repo/image:latest"
        ],
        "RepoDigests": [
            "internal_repo/image@sha256:bc287cd9d506ede8ccd4d5c083ba71b60157b85af90188044860e2cffa5b996d"
        ],
        "Parent": "",
        "Comment": "",
        "Created": "2017-07-05T11:06:34.148937884Z",
        "Container": "d802f466bfefc17dce309a7ec63ea22929377e70f3105f03754d01f298aba457",
        "ContainerConfig": {
            "Hostname": "d802f466bfef",
            "Domainname": "",
            "User": "",
            "AttachStdin": true,
            "AttachStdout": true,
            "AttachStderr": true,
            "Tty": true,
            "OpenStdin": true,
            "StdinOnce": true,
            "Env": null,
            "Cmd": [
                "/bin/bash"
            ],
            "Image": "internal_repo/image",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": {}
        },
        "DockerVersion": "17.05.0-ce",
        "Author": "",
        "Config": {
            "Hostname": "",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "Tty": false,
            "OpenStdin": false,
            "StdinOnce": false,
            "Env": null,
            "Cmd": [
                "/bin/bash"
            ],
            "Image": "",
            "Volumes": null,
            "WorkingDir": "",
            "Entrypoint": null,
            "OnBuild": null,
            "Labels": {}
        },
        "Architecture": "amd64",
        "Os": "linux",
        "Size": 3265786582,
        "VirtualSize": 3265786582,
        "GraphDriver": {
            "Data": null,
            "Name": "aufs"
        },
        "RootFS": {
            "Type": "layers",
            "Layers": [
                "sha256:3cb5cef9e16b6c2fc4c25591b64c3a37ee728e41a9dbb02be96e94ee0784536d",
                "sha256:cf19831f7cead5e80c994c75812f26e8010ad4d422b1693b73cd2b06e8890913",
                "sha256:092bf598ac10f3529cb9225e560c903aaae3ed524e9e793c74b5c1ea71cf26de"
            ]
        }
    }
]
```

The only difference in the output for 17.06 is

```
<         "Size": 3265786582,
<         "VirtualSize": 3265786582,
---
>         "Size": 3258003733,
>         "VirtualSize": 3258003733,
```
I have confirmed that in both the 17.05 and 17.06 cases that the `/var/lib/docker/tmp/GetImageBlob*` files created have identical sha256sums, which means the same layers are being pulled from the repo in both cases.

`/var/log/docker.log` does not show any usual errors or output when extracting in either case. Output from the 17.06 case:

```
time="2017-07-13T09:19:18.115570865Z" level=debug msg="Downloaded 515ece5796b4 to tempfile /mnt/sda1/var/lib/docker/tmp/GetImageBlob669060717" 
time="2017-07-13T09:19:19.312413910Z" level=debug msg="Downloaded c10ec98cceed to tempfile /mnt/sda1/var/lib/docker/tmp/GetImageBlob298385640" 
time="2017-07-13T09:21:35.851669441Z" level=debug msg="Downloaded a697c759d873 to tempfile /mnt/sda1/var/lib/docker/tmp/GetImageBlob421049798" 
time="2017-07-13T09:22:55.763161662Z" level=debug msg="Applied tar sha256:3cb5cef9e16b6c2fc4c25591b64c3a37ee728e41a9dbb02be96e94ee0784536d to b60e75153572e514ebc5fa9f6f4589a3306f42d5bd4b738edba74e5b538e6124
, size: 3250951555" 
time="2017-07-13T09:22:55.841672470Z" level=debug msg="Applied tar sha256:cf19831f7cead5e80c994c75812f26e8010ad4d422b1693b73cd2b06e8890913 to c6fb8d7b5c89bfb515dbc6f7ad1e3fc189319c2fb4e4255370883b186e7af36e
, size: 807" 
time="2017-07-13T09:22:55.978860870Z" level=debug msg="Applied tar sha256:092bf598ac10f3529cb9225e560c903aaae3ed524e9e793c74b5c1ea71cf26de to efb16627adf62001b66723902dcc85b58fc4972ff314120ab18185b187c08e45
, size: 7051371" 
```

When reproducing the issue with overlay2, I can confirm that the `/diff` directory of the `a697c759d873` layer in `/var/lib/docker` contains the incorrectly extracted file paths before starting any containers.

### PR DESCRIPTION
**Pull request by tonistiigi**
fixes #33958
fixes #34092

One possible fix for #34092 , especially for the releases that can't wait for golang update.

https://github.com/golang/go/compare/release-branch.go1.8...tonistiigi:revert-prefix-ignore

@dmcgowan @stevvooe @thaJeztah @tiborvass @andrewhsu

### CODE REVIEW

**Comment by tonistiigi** [comment]
We can also discuss better fixes than reverting but these should probably happen in upstream first.

**Comment by thaJeztah** [comment]
@tonistiigi could you add your description in https://github.com/moby/moby/issues/34092#issuecomment-315231720 as commit message? (probably to the "test" commit, but could be both)

File: pkg/archive/archive_test.go [hunk 1]
```diff
@@ -1203,6 +1203,23 @@ func TestReplaceFileTarWrapper(t *testing.T) {
 	}
 }
 
+func TestPrefixHeaderReadable(t *testing.T) {
```
Comment (thaJeztah) [inline pkg/archive/archive_test.go]: Can you add a short description of what we're testing for here?

File: Dockerfile [hunk 2]
```diff
@@ -107,6 +107,7 @@ RUN set -x \
 # IMPORTANT: If the version of Go is updated, the Windows to Linux CI machines
 #            will need updating, to avoid errors. Ping #docker-maintainers on IRC
 #            with a heads-up.
+# IMPORTANT: When updating this please note that stdlib archive/tar pkg is vendored
```
Comment (thaJeztah) [inline Dockerfile]: This Note should be added to all the Dockerfiles

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Fix and code itself LGTM, but left some comments

**Comment by thaJeztah** [comment]
@tonistiigi guess this one also fixes https://github.com/moby/moby/issues/33958 ?

**Comment by tonistiigi** [comment]
> guess this one also fixes #33958 ?

@thaJeztah yes

**Comment by tonistiigi** [comment]
@thaJeztah Updated

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by dmcgowan** [comment]
LGTM

Let's figure out a better way to handle this vendoring in a follow up. I don't think it is a big issue to create a repo with just `archive/tar` from the golang fork, as long as that repo is only updated from the golang fork and not maintained as a package fork.

**Review by dmcgowan — APPROVED** [review]

**Review by vdemeester — APPROVED** [review]
LGTM 🐯

**Pull request merged**